File: packages/claude-code-toolbox/src/agents/runtime/protocols/parallelFanout.ts
/**
 * Parallel fan-out protocol — dispatch the same task to every agent at once,
 * then have the first listed agent synthesize a merged answer from their peers.
 *
 * Respects `agentTeams.maxConcurrentAgents` via an internal worker-pool so we
 * never blow past the configured concurrency cap (wired in runOrchestrator).
 */
import type { Protocol } from "../runtimeTypes";
import { addUsage, emitMessage, findAgentByName, makeTotals, runOneTurn } from "./shared";

export type ParallelFanoutOptions = {
  maxConcurrent: number;
};

export function makeParallelFanout(opts: ParallelFanoutOptions): Protocol {
  return async (ctx) => {
    const totals = makeTotals();
    const roster = ctx.team.agents
      .map((n) => findAgentByName(ctx.agents, n))
      .filter((a): a is NonNullable<typeof a> => a !== undefined);
    if (!roster.length) {
      return { status: "error", totals };
    }
    const synthesizer = roster[0];
    const results = new Array<string>(roster.length);
    let nextIdx = 0;
    const workers: Promise<void>[] = [];
    const limit = Math.max(1, Math.min(opts.maxConcurrent, roster.length));
    let sharedError = false;

    const workerLoop = async (): Promise<void> => {
      while (!sharedError && !ctx.signal.aborted) {
        const idx = nextIdx++;
        if (idx >= roster.length) {
          return;
        }
        const agent = roster[idx];
        const prompt = [
          `Parallel fan-out: every agent on the team is answering this task independently.`,
          `Team task:`,
          ctx.userPrompt,
          "",
          `You are "${agent.name}" (${agent.role}). Answer from your perspective only. Keep it self-contained.`,
        ].join("\n");
        const localTotals = makeTotals();
        const res = await runOneTurn({
          spawn: ctx.spawnAgentTurn,
          agent,
          prompt,
          runId: ctx.runId,
          turn: idx + 1,
          phase: "none",
          bus: ctx.bus,
          totals: localTotals,
          cwd: ctx.cwd,
          signal: ctx.signal,
          claudeBin: ctx.claudeBin,
        });
        addUsage(totals, localTotals);
        if (res.errored) {
          sharedError = true;
          return;
        }
        results[idx] = res.text;
      }
    };

    for (let i = 0; i < limit; i++) {
      workers.push(workerLoop());
    }
    await Promise.all(workers);

    if (sharedError) {
      return { status: "error", totals };
    }
    if (ctx.signal.aborted) {
      return { status: "aborted", totals };
    }

    const merged = roster
      .map((a, i) => `### ${a.name} (${a.role})\n\n${(results[i] || "(no response)").trim()}`)
      .join("\n\n");
    emitMessage(ctx.bus, "fan-out", synthesizer.name, "all peer responses forwarded for synthesis");
    const synthPrompt = [
      `You are "${synthesizer.name}". Your teammates answered the same task in parallel.`,
      `Synthesize a single merged answer that cites or resolves any disagreements.`,
      "",
      `Team task:`,
      ctx.userPrompt,
      "",
      `Peer responses:`,
      merged,
    ].join("\n");
    const synthTotals = makeTotals();
    const synthRes = await runOneTurn({
      spawn: ctx.spawnAgentTurn,
      agent: synthesizer,
      prompt: synthPrompt,
      runId: ctx.runId,
      turn: roster.length + 1,
      phase: "none",
      bus: ctx.bus,
      totals: synthTotals,
      cwd: ctx.cwd,
      signal: ctx.signal,
      claudeBin: ctx.claudeBin,
    });
    addUsage(totals, synthTotals);
    if (synthRes.errored) {
      return { status: "error", totals };
    }
    return { status: "completed", totals };
  };
}

File: packages/claude-code-toolbox-intellij/src/main/kotlin/com/amitchorasiya/claude/toolbox/intellij/agents/runtime/protocols/ParallelFanout.kt
package com.amitchorasiya.claude.toolbox.intellij.agents.runtime.protocols

import com.amitchorasiya.claude.toolbox.intellij.agents.runtime.*
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.atomic.AtomicInteger

fun parallelFanout(ctx: ProtocolContext, maxConcurrent: Int = 3): ProtocolResult {
    val totals = RunUsage()
    val roster = ctx.agents
    if (roster.isEmpty()) return ProtocolResult("error", totals)

    emitMessage(ctx.bus, "system", "all", "Phase 1: Fan-out — ${roster.size} agents working in parallel")
    ctx.bus.emit(AgentRunEvent.PhaseBoundary(nowIso(), ctx.runId, "none", "none", false))

    val pool = Executors.newFixedThreadPool(maxConcurrent.coerceAtMost(roster.size))
    val turnCounter = AtomicInteger(0)
    val results = mutableListOf<Pair<String, String>>()

    try {
        val futures = mutableListOf<Future<Pair<String, TurnResult>>>()
        for (agent in roster) {
            futures.add(pool.submit<Pair<String, TurnResult>> {
                val turn = turnCounter.incrementAndGet()
                val prompt = "Team task: ${ctx.userPrompt}\n\nYou are \"${agent.name}\". Provide your complete perspective independently."
                val r = spawnAgentTurn(agent, prompt, ctx.runId, turn, "none", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
                agent.name to r
            })
        }
        for (f in futures) {
            val (name, r) = f.get()
            if (r.aborted) return ProtocolResult("aborted", totals)
            if (!r.errored) results.add(name to r.text)
        }
    } finally {
        pool.shutdownNow()
    }

    if (results.isEmpty()) return ProtocolResult("error", totals)
    if (ctx.abortFlag.get()) return ProtocolResult("aborted", totals)

    emitMessage(ctx.bus, "system", "all", "Phase 2: Synthesis — merging ${results.size} responses")
    val judgeAgent = if (ctx.team.judge != null) findAgentByName(roster, ctx.team.judge) ?: roster[0] else roster[0]
    val transcript = describeTranscriptForJudge(results)
    val judgePrompt = "You are the synthesizer \"${judgeAgent.name}\".\n\nTeam task: ${ctx.userPrompt}\n\nAll team responses:\n$transcript\n\nProduce a single cohesive answer incorporating the best insights."
    val turn = turnCounter.incrementAndGet()
    val judgeResult = spawnAgentTurn(judgeAgent, judgePrompt, ctx.runId, turn, "none", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
    if (judgeResult.aborted) return ProtocolResult("aborted", totals)
    return ProtocolResult(if (judgeResult.errored) "error" else "completed", totals)
}
